### PR TITLE
Implement field mask option when fetching document snapshots in bulk

### LINKED ISSUE #2048: FireStore get and batchget with mask
In the firestore low level API there was [mask functionality for filtering the returned document](https://cloud.google.com/firestore/docs/reference/rest/v1beta1/projects.databases.documents/batchGet#request-body) only for data that needed. It useful when we have large document but using specific api that only interest in some field

Could we also have it in high level API?

### PR DESCRIPTION
**Pull request by jskeet**
Fixes #2048 (for batchGet anyway; I don't expect we'll do this for get, where there's significantly less impact).

### CODE REVIEW

File: apis/Google.Cloud.Firestore/Google.Cloud.Firestore.Tests/FirestoreDbTest.cs [hunk 1]
```diff
@@ -283,7 +283,31 @@ public async Task GetDocumentSnapshotsAsync_WithTransaction()
...
             var db = FirestoreDb.Create("proj", "db", mock.Object);
-            var results = await db.GetDocumentSnapshotsAsync(new[] { db.Document("col1/doc1") }, transactionId, default);
+            var results = await db.GetDocumentSnapshotsAsync(new[] { db.Document("col1/doc1") }, transactionId, fieldMask: null, default);
+            mock.VerifyAll();
+        }
+
+        [Fact]
+        public async Task GetDocumentSnapshotsAsync_WithFieldMask()
+        {
+            // This test is about checking the transaction is copied...
```
Comment (amanda-tarafa) [inline apis/Google.Cloud.Firestore/Google.Cloud.Firestore.Tests/FirestoreDbTest.cs]: This comment here is probably from copy/paste.

File: apis/Google.Cloud.Firestore/Google.Cloud.Firestore.Tests/FirestoreDbTest.cs [hunk 2]
```diff
@@ -283,7 +283,31 @@ public async Task GetDocumentSnapshotsAsync_WithTransaction()
...
+            var transactionId = ByteString.CopyFrom(1, 2, 3, 4);
```
Comment (amanda-tarafa) [inline apis/Google.Cloud.Firestore/Google.Cloud.Firestore.Tests/FirestoreDbTest.cs]: And this is not used in the test.

File: apis/Google.Cloud.Firestore/Google.Cloud.Firestore.Tests/ComparisonTester.cs [hunk 3]
```diff
@@ -22,7 +22,10 @@ internal static class ComparisonTester
         internal static void AssertComparison<T>(T smaller, T larger, IComparer<T> comparer = null)
         {
             comparer = comparer ?? Comparer<T>.Default;
-            Assert.InRange(comparer.Compare(smaller, larger), int.MinValue, -1);
+            Assert.True(comparer.Compare(smaller, larger) < 0,
+                $"Compare({smaller}, {larger}) should be < 0");
+            Assert.True(comparer.Compare(larger, smaller) > 0,
+                $"Compare({larger}, {smaller}) should be > 0.");
             Assert.InRange(comparer.Compare(larger, smaller), 1, int.MaxValue);
```
Comment (amanda-tarafa) [inline apis/Google.Cloud.Firestore/Google.Cloud.Firestore.Tests/ComparisonTester.cs]: You can remove this one I think.

Comment (jskeet) [reply apis/Google.Cloud.Firestore/Google.Cloud.Firestore.Tests/ComparisonTester.cs]: Whoops, yes. Done.

File: apis/Google.Cloud.Firestore/Google.Cloud.Firestore.Tests/FirestoreDbTest.cs (hunk 1 above)
Comment (jskeet) [reply apis/Google.Cloud.Firestore/Google.Cloud.Firestore.Tests/FirestoreDbTest.cs]: Yup! Will delete the transaction bits.

File: apis/Google.Cloud.Firestore/Google.Cloud.Firestore/FieldMask.cs [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+        // provides setwise behavior too.
+        private readonly SortedSet<FieldPath> _fieldPaths;
+
+        private FieldMask(IEnumerable<FieldPath> paths)
+        {
+            _fieldPaths = new SortedSet<FieldPath>(paths);
+        }
+
+        /// <summary>
+        /// Creates a path from multiple segments. Each segment is treated verbatim: it may contain dots, which will lead to the segment being escaped in the
```
[the remark this answers is not in the document]
Comment (jskeet) [reply apis/Google.Cloud.Firestore/Google.Cloud.Firestore/FieldMask.cs]: Yup, I need to rewrite both constructor doc strings.

File: apis/Google.Cloud.Firestore/Google.Cloud.Firestore/FieldPath.cs [hunk 5]
```diff
@@ -165,6 +170,33 @@ private static bool IsValidIdentifier(string identifier)
...
+            if (other is null)
+            {
+                return 1;
+            }
+            for (int i = 0; i < Segments.Length; i++)
+            {
+                // If other is a strict prefix of this path, it comes before this path.
+                if (i >= other.Segments.Length)
+                {
+                    return 1;
```
[the remark this answers is not in the document]
Comment (jskeet) [reply apis/Google.Cloud.Firestore/Google.Cloud.Firestore/FieldPath.cs]: Have refactored a bit so it's simpler.

File: apis/Google.Cloud.Firestore/Google.Cloud.Firestore/FieldMask.cs [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+        // provides setwise behavior too.
+        private readonly SortedSet<FieldPath> _fieldPaths;
+
+        private FieldMask(IEnumerable<FieldPath> paths)
+        {
+            _fieldPaths = new SortedSet<FieldPath>(paths);
+        }
+
+        /// <summary>
+        /// Creates a path from the given paths. Each path is treated as a dot-separated sequence of field names.
```
Comment (amanda-tarafa) [inline apis/Google.Cloud.Firestore/Google.Cloud.Firestore/FieldMask.cs]: Creates a mask.

Comment (jskeet) [reply apis/Google.Cloud.Firestore/Google.Cloud.Firestore/FieldMask.cs]: Doh! Thanks. I shouldn't code before coffee. Will fix, rebase, push, wait for green, merge.

**Pull request merged**